File: cpp/97.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  bool isInterleave(string s1, string s2, string s3) {
    int n1 = s1.length(), n2 = s2.length(), n3 = s3.length();
    if(n1 + n2 != n3)
      return false;
    bool dp[n1+1][n2+1];
    for(int i=0; i<=n1; i++)
      for(int j=0; j<=n2; j++){
        if(!i && !j)
          dp[i][j] = true;
        else
          dp[i][j] = (s1[i-1] == s3[i+j-1] && dp[i-1][j]) || (s2[j-1] == s3[i+j-1] && dp[i][j-1]);
        }
    return dp[n1][n2];
  }
};

int main(){
  auto sol = Solution();
  cout << sol.isInterleave("aabcc", "dbbca", "aadbbcbcac");
  cout << sol.isInterleave("aabcc", "dbbca", "aadbbbaccc");
}

File: python/97.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        n1, n2, n3 = len(s1), len(s2), len(s3)
        if n1 + n2 != n3:
            return False
        dp = [[False] * (n2+1) for j in xrange(n1+1)]
        for i in xrange(n1+1):
            for j in xrange(n2+1):
                if i == j == 0:
                    dp[0][0] = True
                else:
                    dp[i][j] = (i > 0 and s3[i+j-1] == s1[i-1] and dp[i-1][j]) or\
                               (j > 0 and s3[i+j-1] == s2[j-1] and dp[i][j-1])
        return dp[n1][n2]
